File: android/milestone2kt/src/main/java/dev/kaya/milestone2kt/Reorder.java
package dev.kaya.milestone2kt;

import dev.kaya.KayaApp;
import dev.kaya.KayaRecords;

import java.util.List;

/**
 * The reorder scene from the JVM: order as collection data, end to
 * end. Three stamped rows and two buttons that never touch a widget —
 * each handler repositions an entry by key (collection_move on the
 * wire, move_child at the toolkit), and the selftest's expect_order
 * reads the toolkit's actual child order back. The root is a row so
 * the For's container is the scene's only column-kind widget:
 * languages disagree on whether containers are created before or
 * after their children, and column#0 must name the same widget
 * everywhere.
 */
final class Reorder {
    /** The record is the schema. */
    record Item(String title) {}

    static void app() {
        KayaApp app = new KayaApp();

        app.build(tx -> {
            KayaRecords.Collection<String, Item> items =
                    KayaRecords.collectionOf(tx, Item.class);

            tx.mount(tx.row(
                    tx.button("rotate", t -> {
                        // First entry to the end. The model owns the
                        // order, so the handler asks it which key is
                        // first — it never counts widgets.
                        List<KayaRecords.Entry<String, Item>> entries = items.items(t);
                        items.moveToEnd(t, entries.get(0).key);
                    }),
                    tx.button("lift", t -> {
                        // Last entry to the front: moveToFront is
                        // sugar for moveBefore the current first key —
                        // the same wire op, keys never indices.
                        List<KayaRecords.Entry<String, Item>> entries = items.items(t);
                        items.moveToFront(t, entries.get(entries.size() - 1).key);
                    }),
                    tx.forEach(items.handle, t -> {
                        items.label(t, Item::title);
                    })));
            for (String key : new String[] { "a", "b", "c" }) {
                items.insert(tx, key, new Item(key));
            }
            return null;
        });

        app.dispatchLoop();
    }

    private Reorder() {}
}

File: crates/kaya/examples/reorder.rs
//! The reorder scene: order as collection data, end to end. Three
//! stamped rows and two buttons that never touch a widget — each
//! handler repositions an entry by key (collection_move on the wire,
//! move_child at the toolkit), and the selftest's expect_order reads
//! the toolkit's actual child order back, which no creation-ordered
//! registry could observe.
//!
//! The backend selftest (KAYA_SELFTEST=reorder) checks "a|b|c", clicks
//! rotate (first entry to the end), checks "b|c|a", clicks lift (last
//! entry before the first), and checks "a|b|c" again.

use kaya::Occurrence;

kaya::record! {
    struct Item {
        title: String,
    }
}

pub(crate) fn app(ctx: kaya::AppCtx) {
    let mut tx = ctx.begin();
    let items = tx.collection::<Item>();
    let rotate = tx.button("rotate");
    let lift = tx.button("lift");
    let (list, ()) = tx.for_each(&items, |t| {
        t.label(Item::title());
    });
    let root = tx.column(&[rotate, lift, list]);
    tx.mount(root);
    for key in ["a", "b", "c"] {
        tx.insert(&items, key, Item { title: key.to_string() });
    }
    tx.commit();

    loop {
        match ctx.next() {
            Occurrence::ButtonClicked { id } if id == rotate => {
                // First entry to the end. The model owns the order, so
                // the handler asks it which key is first — it never
                // counts widgets.
                let mut tx = ctx.begin();
                let entries = tx.items(&items);
                let (first, _) = entries.first().expect("reorder scene has entries").clone();
                tx.move_to_end(&items, first);
                tx.commit();
            }
            Occurrence::ButtonClicked { id } if id == lift => {
                // Last entry to the front, anchored by key: keys,
                // never indices.
                let mut tx = ctx.begin();
                let entries = tx.items(&items);
                let (last, _) = entries.last().expect("reorder scene has entries").clone();
                let (first, _) = entries.first().expect("reorder scene has entries").clone();
                tx.move_before(&items, last, first);
                tx.commit();
            }
            Occurrence::Shutdown => break,
            _ => {}
        }
    }
}

fn main() {
    kaya::run(app)
}
